src/analysis/processing/qgsalgorithmextractlabels.cpp

```diff
@@ -0,0 +1,464 @@
+/***************************************************************************
+  qgsalgorithmextractlabels.cpp - QgsExtractLabelsAlgorithm
+ ---------------------
+ begin                : 30.12.2021
+ copyright            : (C) 2021 by Mathieu Pellerin
+ email                : nirvn dot asia at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsalgorithmextractlabels.h"
+#include "qgsprocessingparameters.h"
+#include "qgsmapthemecollection.h"
+#include "qgsmaprenderercustompainterjob.h"
+#include "qgsnullpainterdevice.h"
+#include "qgslabelsink.h"
+#include "qgslayertree.h"
+#include "qgsvectorlayer.h"
+#include "qgsscalecalculator.h"
+#include "qgstextlabelfeature.h"
+#include "qgsnullsymbolrenderer.h"
+
+#include "pal/feature.h"
+#include "pal/pointset.h"
+#include "pal/labelposition.h"
+
+#include <QPainter>
+
+#include <cmath>
+
+///@cond PRIVATE
+
+QString QgsExtractLabelsAlgorithm::name() const
+{
+  return QStringLiteral( "extractlabels" );
+}
+
+QString QgsExtractLabelsAlgorithm::displayName() const
+{
+  return QObject::tr( "Extract labels" );
+}
+
+QStringList QgsExtractLabelsAlgorithm::tags() const
+{
+  return QObject::tr( "map themes,font,position" ).split( ',' );
+}
+
+QgsProcessingAlgorithm::Flags QgsExtractLabelsAlgorithm::flags() const
+{
+  return QgsProcessingAlgorithm::flags() | FlagRequiresProject;
+}
+
+QString QgsExtractLabelsAlgorithm::group() const
+{
+  return QObject::tr( "Cartography" );
+}
+
+QString QgsExtractLabelsAlgorithm::groupId() const
+{
+  return QStringLiteral( "cartography" );
+}
+
+void QgsExtractLabelsAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterExtent(
+                  QStringLiteral( "EXTENT" ),
+                  QObject::tr( "Map extent" ) ) );
+
+  addParameter( new QgsProcessingParameterScale(
+                  QStringLiteral( "SCALE" ),
+                  QObject::tr( "Map scale" ) ) );
+
+  std::unique_ptr<QgsProcessingParameterMapTheme> mapThemeParameter = std::make_unique<QgsProcessingParameterMapTheme>(
+        QStringLiteral( "MAP_THEME" ),
+        QObject::tr( "Map theme" ),
+        QVariant(), true );
+  mapThemeParameter->setHelp( QObject::tr( "This parameter is optional. When left unset, the algorithm will fallback to match the currently opened project's layer tree visibility to extract labels from." ) );
```

Comment:
```suggestion
  mapThemeParameter->setHelp( QObject::tr( "This parameter is optional. When left unset, the algorithm will fallback to extracting labels from all currently visible layers in the project." ) );
```

"layer tree" is an internal name, we never expose this to users